File: src/main/java/vergisst/minecraftmod/weaponthrow/handler/PacketHandler.java
package vergisst.minecraftmod.weaponthrow.handler;

import net.fabricmc.fabric.api.networking.v1.PlayerLookup;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.entity.Entity;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.server.world.ServerWorld;
import vergisst.minecraftmod.weaponthrow.packets.BasePacket;
import vergisst.minecraftmod.weaponthrow.packets.CPacketThrow;


public class PacketHandler {
    public static void registerServerListener() {
        CPacketThrow.register();
    }

    public static void sendToAllTracking(Entity entity, BasePacket packet) {
        if (entity.getWorld() instanceof ServerWorld) {
            // back to player
            if (entity instanceof ServerPlayerEntity self) {
                ServerPlayNetworking.send(self, packet.getIdentifier(), packet.getBuf());
            }
            // all the player tracing entity
            for (ServerPlayerEntity player : PlayerLookup.tracking((ServerWorld) entity.getWorld(), entity.getBlockPos())) {
                ServerPlayNetworking.send(player, packet.getIdentifier(), packet.getBuf());
            }
        }
    }

    public static void sendToAll(Entity entity, BasePacket packet) {
        for (ServerPlayerEntity player : PlayerLookup.tracking((ServerWorld) entity.getWorld(), entity.getBlockPos())) {
            ServerPlayNetworking.send(player, packet.getIdentifier(), packet.getBuf());
        }
    }
}

File: src/main/kotlin/vergisst/minecraftmod/weaponthrowlite/handlers/PacketHandler.kt
package vergisst.minecraftmod.weaponthrowlite.handlers

import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.fabricmc.fabric.api.networking.v1.PlayerLookup
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.entity.Entity
import net.minecraft.server.world.ServerWorld
import vergisst.minecraftmod.weaponthrowlite.packets.BasePacket
import vergisst.minecraftmod.weaponthrowlite.packets.CPacketThrow

object PacketHandler {

    fun registerServerListeners() {
        CPacketThrow.register()
    }

    fun sendToServer(packet: BasePacket) {
        ClientPlayNetworking.send(packet.identifier, packet.buf)
    }

    fun sendToAll(entity: Entity, packet: BasePacket) {
        for (player in PlayerLookup.tracking(entity.world as ServerWorld, entity.blockPos)) {
            ServerPlayNetworking.send(player, packet.identifier, packet.buf)
        }
    }
}